### ISSUE TITLE
distributed.Client.processing method hangs when called from async code 

### ISSUE THREAD

**Issue by hdoupe** [dask/distributed#1952]
I'm tinkering around with building a [REST api in front of a dask cluster](https://github.com/hdoupe/async-dask-api). I'd like to be able to query the dask scheduler to see which jobs are still running. However, the `distributed.Client.processing` method appears to hang when called from asynchronous code. It appears to work just fine with synchronous code. I have a simple tornado application that demonstrates this:

python versoin: 3.6.5
tornado version: 5.0.2
distributed version: 1.21.6
dask version: 0.17.2

```
import json

from tornado.web import Application, RequestHandler
from tornado.ioloop import IOLoop

from distributed import Client

class AsyncTaskStatus(RequestHandler):

    async def get(self):
        print('GET')
        client = await Client(asynchronous=True)
        print('got client', client)
        print('client', client._scheduler_identity)

        # gets hung up here
        processing = await client.processing()

        print('processing tasks', processing)

        self.write(json.dumps({addr: list(processing[addr]) for addr in processing}))


class SyncTaskStatus(RequestHandler):

    def get(self):
        print('GET')
        client = Client()
        print('got client', client)
        print('client', client._scheduler_identity)

        # works just fine
        processing = client.processing()

        print('processing tasks', processing)

        self.write(json.dumps({addr: list(processing[addr]) for addr in processing}))


class Health(RequestHandler):

    async def get(self):
        print('GET-HEALTH')
        self.write('healthy...')


def make_app():
    return Application(
        [(r'/asynctaskstatus', AsyncTaskStatus),
         (r'/synctaskstatus', SyncTaskStatus),
         (r'/health', Health)],
        default_host='0.0.0.0',
        debug=True,
        autoreload=True
    )

if __name__ == "__main__":
    print('starting up app...')
    app = make_app()
    app.listen(8888)
    IOLoop.current().start()
```

API calls:
```
HDoupe-MacBook-Pro:async-dask-api henrydoupe$ http http://localhost:8888/synctaskstatus
HTTP/1.1 200 OK
Content-Length: 116
Content-Type: text/html; charset=UTF-8
Date: Wed, 02 May 2018 21:35:50 GMT
Etag: "51ddcc8412a5a82c61422a58b84aab85d7f847e6"
Server: TornadoServer/5.0.2

{
    "tcp://127.0.0.1:55274": [],
    "tcp://127.0.0.1:55278": [],
    "tcp://127.0.0.1:55280": [],
    "tcp://127.0.0.1:55282": []
}

HDoupe-MacBook-Pro:async-dask-api henrydoupe$ http http://localhost:8888/asynctaskstatus

http: error: Request timed out (30s).
HDoupe-MacBook-Pro:async-dask-api henrydoupe$ 
```

App output:
```
HDoupe-MacBook-Pro:dask_processing_ex henrydoupe$ python dist_ex.py 
starting up app...
GET
got client <Client: scheduler='tcp://127.0.0.1:55268' processes=4 cores=4>
client {'type': 'Scheduler', 'id': 'Scheduler-82dbce28-eee2-4f2f-adff-886729f7da7c', 'address': 'tcp://127.0.0.1:55268', 'services': {'bokeh': 55269}, 'workers': {'tcp://127.0.0.1:55274': {'name': 'tcp://127.0.0.1:55274', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55272, 'bokeh': 55275}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-35fo0m15', 'pid': 15420, 'cpu': 0.0, 'memory': 66322432, 'time': 1525296953.562387, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55278': {'name': 'tcp://127.0.0.1:55278', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55270, 'bokeh': 55279}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-t2g65k71', 'pid': 15418, 'cpu': 0.0, 'memory': 66461696, 'time': 1525296953.613239, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55280': {'name': 'tcp://127.0.0.1:55280', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55271, 'bokeh': 55281}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-ln_9mof8', 'pid': 15419, 'cpu': 0.0, 'memory': 66367488, 'time': 1525296953.6248488, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55282': {'name': 'tcp://127.0.0.1:55282', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55273, 'bokeh': 55283}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-hwhailqr', 'pid': 15421, 'cpu': 0.0, 'memory': 66506752, 'time': 1525296953.62933, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}}}
processing tasks {'tcp://127.0.0.1:55274': set(), 'tcp://127.0.0.1:55278': set(), 'tcp://127.0.0.1:55280': set(), 'tcp://127.0.0.1:55282': set()}
GET
got client <Client: scheduler='tcp://127.0.0.1:55296' processes=4 cores=4>
client {'type': 'Scheduler', 'id': 'Scheduler-6ab99b04-c221-4a17-b303-05cfac4259c0', 'address': 'tcp://127.0.0.1:55296', 'services': {'bokeh': 55297}, 'workers': {'tcp://127.0.0.1:55302': {'name': 'tcp://127.0.0.1:55302', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55301, 'bokeh': 55303}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-i1u8iesj', 'pid': 15430, 'cpu': 0.0, 'memory': 66334720, 'time': 1525296959.923568, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55304': {'name': 'tcp://127.0.0.1:55304', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55300, 'bokeh': 55305}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-kefmgt07', 'pid': 15429, 'cpu': 0.0, 'memory': 66453504, 'time': 1525296959.941612, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55307': {'name': 'tcp://127.0.0.1:55307', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55298, 'bokeh': 55308}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-f1sd_acw', 'pid': 15427, 'cpu': 0.0, 'memory': 66412544, 'time': 1525296959.953994, 'read_bytes': 0.0, 'write_bytes': 0.0, 'num_fds': 27}, 'tcp://127.0.0.1:55314': {'name': 'tcp://127.0.0.1:55314', 'memory_limit': 4294967296, 'host': '127.0.0.1', 'resources': {}, 'ncores': 1, 'services': {'nanny': 55299, 'bokeh': 55315}, 'local_directory': '/Users/henrydoupe/Documents/async-dask-api/dask_processing_ex/dask-worker-space/worker-fr59jtoq', 'pid': 15428, 'cpu': 0.0, 'memory': 66433024, 'time': 1525296959.98482, 'read_bytes': 2137863.263315082, 'write_bytes': 2137863.263315082, 'num_fds': 27}}}

```

The same behavior is shown on a google cloud cluster that is set up as suggested by the [dask/helm-chart](https://github.com/dask/helm-chart) repo.

My questions are:
1. Am I doing something wrong?
2. Is this desired behavior?
3. If this is desired behavior, is there another method that delivers the same information and can be called from asynchronous code?

Thanks in advance for taking a look at this!

**Comment by hdoupe** [comment]
@mrocklin Thanks for your response. 

Ok, I see. Losing the `set` output type makes sense, especially since similar methods like `has_what` return a dictionary of lists.

Sure, I'll give it a shot. Should I apply this same fix to `Client.stacks`, too? Also, do you have any tests in mind that I could add to pin this behavior down?

**Comment by mrocklin** [comment]
You could mimic the current tests/test_client.py::test_get_processing_sync
test, but using gen_cluster (hopefully the other gen_cluster tests give
good examples on how these work).  Also see
http://distributed.readthedocs.io/en/latest/develop.html#writing-tests

**Comment by hdoupe** [comment]
Ok, thanks @mrocklin. I'll try to open a PR by the end of the upcoming weekend.

**Comment by hdoupe** [comment]
Closed via #1962
